### PR TITLE
Partial rechunks within P2P

### LINKED ISSUE #8326: P2P Rechunking graph prohibits culling and partial release
The P2P rechunk graph is similar to the shuffle graph one large globally connected graph. However, many rechunk applications only operate on slices of the data and is often not connecting all in- and output partitions but only planes in the dataset.

Having only slices/planes connected by the graph is beneficial since it allows us to cull the graph, process downstream releaser earlier and release disk spaces sooner.

A real life example showing this effect is using the public ERA5 dataset (see code below) but a simpler variant can be constructed using artificial data.

<details>
<summary>ERA5 example</summary>

```python
import itertools

import coiled
import dask
import distributed
import flox.xarray
import fsspec
import numpy as np
import pandas as pd
import xarray as xr
fs = fsspec.filesystem("s3")
prefix = "s3://era5-pds/zarr"
variables = [
    store.split("/")[-1] for store in fs.glob("era5-pds/zarr/1979/01/data/*.zarr")
]
years = [path.split("/")[-1] for path in fs.glob("era5-pds/zarr/*")]
months = [f"{m:02d}" for m in range(1, 13)]
last_months = [path.split("/")[-1] for path in fs.glob(f"era5-pds/zarr/{years[-1]}/*")]
var = ("precipitation_amount_1hour_Accumulation.zarr",)
var = (variables[0],)
all_stores = [
    "/".join(t)
    for t in itertools.chain(
        itertools.product((prefix,), years[:-1], months, ("data",), var),
        itertools.product((prefix,), years[-2:], last_months, ("data",), var),
    )
]

def preprocess(ds):
    """Edit the dataset so it combines nicely."""
    (time_dim,) = [dim for dim in ds.dims if "time" in dim]
    ds = ds.rename({time_dim: "time"})
    bounds_var = [var for var in ds.data_vars if "bounds" in var]
    ds = ds.drop_vars(bounds_var)
    return ds

ds = xr.open_mfdataset(
    all_stores,
    engine="zarr",
    combine="nested",
    concat_dim="time",
    preprocess=preprocess,
    join="override",
    parallel=True,
)

import dask
dask.config.set({"array.rechunk.method": "p2p"})

def rechunk_to_frequency(ds, **kwargs):
    newchunks = {}
    for dim, freq in kwargs.items():
        newchunks[dim] = tuple(
            ds[dim]
            .copy(data=np.ones(ds[dim].shape, dtype=np.int64))
            .to_dataframe()
            .resample(freq)
            .sum()[dim]
            .values
        )
    return newchunks

newchunks = rechunk_to_frequency(ds, time="A")
rechunked = ds.chunk(**newchunks, lat=75, lon=75)

# Run any kind of reduction on this. This reduces per year
# so every reduction step should only load 1/45 of the data
result = flox.xarray.xarray_reduce(
    rechunked,
    rechunked.time.dt.year,
    func="quantile",
    skipna=False,
    q=0.9,
    method="blockwise",
)

# Even though we're selecting only a thin slices, this is loading the entire data
# since culling is not possible. The entire data also has to be stored to disk temporarily
# such that the cluster has to be really large for this to work even though the actual
# slices should only be a couple GBs
result.isel(year=slice(2)).compute()
```

</details>

A simpler, toy example showing the same principle is just concatenating a single dimension.
```python

import dask.array as da
# This is a simple example where we're just compacting the first dimension
# The ERA5 example simultaneously also splits another dimension but this boils
# down to the same principle
arr = da.random.random((10, 45), chunks=(1, 1))
rechunked = arr.rechunk((-1, 1))
dask.optimize(rechunked[:, :, :2])[0]
```

```python
import dask
import dask.array as da

arr = da.random.random((10, 45), chunks=(1, 1))
for method in ["p2p", "tasks"]:
    dask.config.set({"array.rechunk.method": method})
    rechunked = arr.rechunk((-1, 1))
    
    len_with_slicing = len(dask.optimize(rechunked[:, :2])[0].__dask_graph__())
    len_without_slicing = len(dask.optimize(rechunked)[0].__dask_graph__())
    print(f"Tasks after slicing\t{method=}\t{len_with_slicing/len_without_slicing*100:2.0f}%")
```

```python
Tasks after slicing	method='p2p'	96%
Tasks after slicing	method='tasks'	 5%
```

For P2P to do the same, we should instead of generating one large P2P layer generate many small P2P layers such that we only connect the relevant slices.

### PR DESCRIPTION
**Pull request by hendrikmakait**
Closes #8326 
Blocked by #8207

~This PR currently splits on _entirely_ independent subset of the rechunking. A more intricate (yet better) implementation would create independent shuffles even for cases where an input would go into more than one shuffle.~

This PR splits any outputs that allows us to cull more inputs, i.e., two outputs along an axis are separated if they do not end in the same input chunk.

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Comment by hendrikmakait** [comment]
There's some stuff we still calculate for the individual n-dimensional partials (scaling as a product) but that could be computed for the partials per axis (scaling as a sum). I'll take a another look at some quick refactoring here but this might very well be premature optimization.

File: distributed/shuffle/_rechunk.py [hunk 1]
```diff
@@ -490,8 +755,16 @@ def create_run_on_worker(
 
 
 def _get_worker_for_hash_sharding(
-    output_partition: NDIndex, workers: Sequence[str]
+    new: ChunkedAxes, partition: NDIndex, workers: Sequence[str]
```
Comment (hendrikmakait) [inline distributed/shuffle/_rechunk.py:758]: FIXME: This is a hacked version of #8207, refactor or move to different PR.

**Comment by hendrikmakait** [comment]
When including the functional changes from #8207, the performance on this PR looks promising, we trade better runtime for increased memory consumption in `tiles_to_rows` which conceptually benefits from being able to split the rechunk into multiple parts because it allows it for better pipelining:

**Wall clock**
Screenshot 2023-11-29 at 15 26 50

**Average memory**
Screenshot 2023-11-29 at 15 26 57

**Peak memory**
Screenshot 2023-11-29 at 15 27 04


The impact of these changes on `test_swap_axes` shown in the benchmarks is interesting. Conceptuall, this test should be unaffected. Indeed, when running this test in isolation, it doesn't show any changes.
Screenshot 2023-11-29 at 15 30 46


Screenshot 2023-11-29 at 15 30 54

File: distributed/shuffle/_rechunk.py [hunk 2]
```diff
@@ -166,45 +169,310 @@ def rechunk_unpack(
...
+
+    partial_axes = []
+    for axis_index, slices in enumerate(sliced_axes):
+        partials = []
+        for slice_ in slices:
+            last_old_chunk: int
+            first_old_chunk, first_old_slice = _old_to_new[axis_index][slice_.start][0]
+            last_old_chunk, last_old_slice = _old_to_new[axis_index][slice_.stop - 1][
+                -1
+            ]
```
Comment (crusaderky) [inline distributed/shuffle/_rechunk.py:265]: Slightly more pythonic and coherent with `_partial_slices`:
```suggestion
    for slices, old_to_new_axis in zip(sliced_axes, _old_to_new):
        partials = []
        for slice_ in slices:
            last_old_chunk: int
            first_old_chunk, first_old_slice = old_to_new_axis[slice_.start][0]
            last_old_chunk, last_old_slice = old_to_new_axis[slice_.stop - 1][-1]
```

File: distributed/shuffle/_rechunk.py [hunk 3]
```diff
@@ -166,45 +169,310 @@ def rechunk_unpack(
...
+        recipe: list[tuple[int, slice]]
+        for new_chunk_index, recipe in enumerate(old_to_new_axis):
+            if len(recipe) == 0:
+                continue
+            current_last_old_chunk, old_slice = recipe[-1]
+            if last_old_chunk is None:
+                last_old_chunk = current_last_old_chunk
+            elif last_old_chunk != current_last_old_chunk:
+                partial_splits.append(new_chunk_index)
+                last_old_chunk = current_last_old_chunk
```
Comment (crusaderky) [inline distributed/shuffle/_rechunk.py:300]:
```suggestion
            if last_old_chunk not in (None, current_last_old_chunk):
                partial_splits.append(new_chunk_index)
            last_old_chunk = current_last_old_chunk
```

File: distributed/shuffle/_rechunk.py [hunk 4]
```diff
@@ -166,45 +169,310 @@ def rechunk_unpack(
...
+    x: da.Array,
+    chunks: ChunkedAxes,
+    ndpartial: _NDPartial,
+    name: str,
+) -> Any:
+    import numpy as np
+
+    from dask.array.chunk import getitem
+
+    dsk: dict = {}
```
Comment (crusaderky) [inline distributed/shuffle/_rechunk.py:328]:
```suggestion
    dsk: dict[Key, Any] = {}
```

File: distributed/shuffle/_rechunk.py [hunk 5]
```diff
@@ -166,45 +169,310 @@ def rechunk_unpack(
...
+            rechunk_transfer,
+            input_task,
+            partial_token,
+            partial_index,
+            partial_new,
+            partial_old,
+            disk,
+        )
 
-    with dask.annotate(shuffle=lambda key: key[1:]):
```
Comment (crusaderky) [inline distributed/shuffle/_rechunk.py:203]: Did this annotation do anything?

File: distributed/shuffle/_rechunk.py [hunk 6]
```diff
@@ -166,45 +169,310 @@ def rechunk_unpack(
...
+        # if B is fully included in the right-most input chunk of A, i.e.,
+        # separating A and B would not allow us to cull more input tasks.
+        last_old_chunk: int | None = (
+            None  # Index of the last input chunk of this partial rechunk
+        )
+        partial_splits = [0]
+        recipe: list[tuple[int, slice]]
+        for new_chunk_index, recipe in enumerate(old_to_new_axis):
+            if len(recipe) == 0:
+                continue
```
Comment (crusaderky) [inline distributed/shuffle/_rechunk.py:294]: coverage says this special case is not covered by unit tests

**Comment by crusaderky** [comment]
Regarding the A/B tests:
`test_tiles_to_rows[0.1-8MiB-p2p-memory]` shows a drastic reduction in memory usage, as expected.
I can't understand why the same on 128 MiB chunks shows a slight increase though?

File: distributed/shuffle/_rechunk.py (hunk 5 above)
Comment (hendrikmakait) [reply distributed/shuffle/_rechunk.py:203]: Nope, I forgot to remove it in #8329.

File: distributed/shuffle/_rechunk.py [hunk 7]
```diff
@@ -166,45 +169,310 @@ def rechunk_unpack(
         )
 
 
-def rechunk_p2p(x: da.Array, chunks: ChunkedAxes) -> da.Array:
-    import numpy as np
+class _Partial(NamedTuple):
```
[the remark this answers is not in the document]
Comment (hendrikmakait) [reply distributed/shuffle/_rechunk.py:172]: Works for me, the intention was to keep as much private as possible, there's no reason for these to be public.

File: distributed/shuffle/_rechunk.py [hunk 8]
```diff
@@ -166,45 +169,310 @@ def rechunk_unpack(
...
+
+class _NDPartial(NamedTuple):
+    """Information used to perform a partial rechunk along all axes"""
+
+    #: n-dimensional slice of the old chunks along each axis that belong to the partial
+    old: NDSlice
+    #: n-dimensional slice of the new chunks along each axis that belong to the partial
+    new: NDSlice
+    #: Indices of the first value of the left-most old chunk along each axis
+    #: to include in this partial. Everything left to this index belongs to
```
[the remark this answers is not in the document]
Comment (hendrikmakait) [reply distributed/shuffle/_rechunk.py:197]: I'll add a few more unit tests around the chunking/partial logic to make this more explicit, this is not straightforward. Going over this piece of code again, I _think_ I can remove this bit of complexity by restructuring some of the logic. So far I've put this off as premature optimization, but it looks like it would also reduce the cognitive load.

**Comment by crusaderky** [comment]
There are a few regressions:
- linting (trivial)
- 10 out of 12 runs failed: test_no_unnecessary_imports_on_worker[scipy] (distributed.tests.test_nanny)
- readthedocs: `ImportError: No module named 'numpy'`

**Comment by crusaderky** [comment]
I've added a few benchmarks: https://github.com/coiled/benchmarks/pull/1236

I'm omitting the null hypothesis, which demonstrates there's virtually zero noise.
I've renamed `test_partial_rechunk` in the screenshots below to `test_adjacent_groups` in the PR above.
A/B test outcome is, overall, quite extreme - in both directions.

image
image

# Observations
- In full O(N^2) problems (`test_swap_axes`) this PR is inconsequential, as expected.
- When p2p rechunk is completely useless - such as in a trivial N:1 aggregation (`test_heal_oversplit`), this PR effectively disables p2p rechunk and falls back to tasks, producing a spectacular reduction in both runtime and memory usage (if you were using in-memory p2p)

_tasks vs p2p, before_
image
_and after_
image


- In localized use cases where the multiple barriers kick in (`test_partial_rechunk` a.k.a. `test_adjacent_groups`, as well as `test_tiles_to_rows`), we see a dramatic reduction in memory usage for in-memory p2p - as expected. However, we also see an equally extreme slowdown when disk access is involved. Notably, there is almost no impact<sup>1</sup>  to the runtime of in-memory p2p (as expected); this highlights how the slowdown is specifically caused by disk access.

<sup>1</sup> There is one regression for in-memory p2p for 8 MiB chunks, which I don't think we should care about.

image
image
image
image



# Conclusions
I don't recommend merging this PR in its current state, as it causes major regressions in several important use cases.
We should understand what changed in the disk access pattern and see if we can remediate.

**Comment by crusaderky** [comment]
test_tiles_to_rows[1-128 MiB-p2p-disk] shows no runtime regression in @hendrikmakait's own A/B tests shown above, while it shows an extreme slowdown in my tests. Any idea what changed?

**Comment by hendrikmakait** [comment]
Not sure, if I had to guess I'd blame recent changes to dask.order. There's not much else that changed since my latest A/B tests. 🤔

**Comment by crusaderky** [comment]
> Not sure, if I had to guess I'd blame recent changes to dask.order. There's not much else that changed since my latest A/B tests. 🤔

Was your test vs. dask/dask git tip as of Nov 29th?

**Comment by hendrikmakait** [comment]
> Was your test vs. dask/dask git tip as of Nov 29th?

Yes.

I wouldn't be surprised if this were the same performance regression we've started seeing in coiled/benchmarks#1228.

**Comment by crusaderky** [comment]
I can't reproduce your A/B results.

- **baseline:** everything main as of Dec 18th
- **1_latest:** dask/dask and coiled/benchmarks main Dec 18th; dask/distributed #8330 Dec 18th
- **4_dependencies:** dask/dask main Nov 29th; coiled/benchmarks main Nov 29th; dask/distributed #8330 Nov 29th

image

**Comment by hendrikmakait** [comment]
@crusaderky: Looking at my old benchmark results, you are in fact reproducing them. However, `test_tiles_to_rows` has significantly improved in performance since then: 

Screenshot 2023-12-18 at 15 28 14

Screenshot 2023-12-18 at 15 29 14

**Comment by hendrikmakait** [comment]
Searching through the commit history, this is likely because https://github.com/dask/distributed/pull/8207 has been merged, which this branch also included at the time.

**Comment by crusaderky** [comment]
The side-by-side dashboard makes it pretty obvious why there's such a dramatic slowdown.

`test_adjacent_groups[0.3-128 MiB-p2p-disk]`
This PR picks for the 180 output chunks even workers only! (worker 0 is always at the bottom)


https://github.com/dask/distributed/assets/6213168/f6077b01-17e3-47c8-86f7-5b00e570ccf6

**Comment by hendrikmakait** [comment]
Quick summary of an offline discussion I had with @crusaderky: In the example above, each partial rechunk has five output tasks (180 inputs / 36 barriers). As a result, the deterministic worker assignment logic will always pick the same five workers to store those outputs which underutilizes the cluster. We should be able to assess the impact of better worker assignment by randomizing it.

**Comment by crusaderky** [comment]
Adding a naive randomization of the workers (https://github.com/dask/distributed/pull/8330/commits/d9116fb4aaaa3209618f2ac641fac0f4464b8b3d) fixes _most_ problems:

image
image

- there are runtime regressions in some tests with 8 MiB chunks. I'm oriented towards accepting them for the greater good.
- there is a runtime regression in `test_swap_axes[1-128 MiB-p2p-disk]`. It makes no sense to me as it should not be impacted by this PR. Investigating.
- the peak memory regressions are false positives, please ignore: https://github.com/coiled/benchmarks/issues/1240
- worker randomization breaks TPCH query 7: https://github.com/dask/distributed/issues/8424

**Comment by hendrikmakait** [comment]
Regarding the regression of `test_swap_axes`, I've seen that happen before when running benchmarks for this PR. In the past, they disappeared if I test that one in isolation. I'm not sure what's going on exactly.

**Comment by crusaderky** [comment]
> Regarding the regression of `test_swap_axes`, I've seen that happen before when running benchmarks for this PR. In the past, they disappeared if I test that one in isolation. I'm not sure what's going on exactly.

Reran it in isolation. You're right, these results don't make sense for me

image

**Comment by hendrikmakait** [comment]
I think what @fjetter and I both have in mind would look like this: https://github.com/hendrikmakait/distributed/pull/14

FWIW, I'd not block on that, a follow-up with another benchmark run is trivial.

**Comment by crusaderky** [comment]
Running a final A/B test. Results out tomorrow morning.

**Comment by crusaderky** [comment]
Looks good. Comment is the same as in the previous posts.

image
image

**Comment by fjetter** [comment]
> I think what @fjetter and I both have in mind would look like this: https://github.com/hendrikmakait/distributed/pull/14

essentially, yes. I would've not shifted by one but by `len(workers)` but the gist is the same. I think shifting by `len(workers)` generates more homogeneous distributions for small numbers of shuffles

**Comment by hendrikmakait** [comment]
> essentially, yes. I would've not shifted by one but by `len(workers)` but the gist is the same. I think shifting by `len(workers)` generates more homogeneous distributions for small numbers of shuffles

I'm confused, wouldn't shifting by `len(workers)` have no effect whatsoever in a cluster without fluctuating workers?

The problem the shift solves is that static range partitioning if `len(partitions) < `len(workers)` skips some workers (e.g., only picks every other worker if `len(partitions) / len(workers) = 0.5`.

**Comment by hendrikmakait** [comment]
Note that we could potentially investigate switching assignment to round-robin and then shift by the `len(partitions) % len(workers)`. IIRC, the lack of buffering in our buffers currently prohibits that though.

**Review by crusaderky — APPROVED** [review]

**Comment by crusaderky** [comment]
:partying_face:
